### PR TITLE
Allow inline json for XCS callbacks. 

### PR DESCRIPTION
**Pull request by nicolaslara**
## What is the purpose of the change

This simplifies the interface of XCS by allowing the `next_memo` key in crosschain swaps and the callbacks key in the outposts to include json directly instead of a string of serialized json. This makes writing those messages easier and avoids growth in the size of the payload when many messages are nested. 

## Brief Changelog

 - next_memo and callbacks are now a json object instead of a string with a serialized json object.

## Testing and Verifying

all tests have been adapted and pass

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: cosmwasm/contracts/crosschain-swaps/src/execute.rs [hunk 1]
```diff
@@ -51,10 +51,8 @@ pub fn swap_and_forward(
     // If there is a memo, check that it is valid (i.e. a valud json object that
     // doesn't contain the key that we will insert later)
     if let Some(memo) = &next_memo {
-        // Parse the string as valid json
-        let json = parse_json(memo)?;
         // Ensure the json is an object ({...}) and that it does not contain the CALLBACK_KEY
-        ensure_key_missing(&json, CALLBACK_KEY)?;
+        ensure_key_missing(memo.as_value(), CALLBACK_KEY)?;
     }
```
Comment (ValarDragon) [inline cosmwasm/contracts/crosschain-swaps/src/execute.rs:56]: Nice!

File: cosmwasm/contracts/crosschain-swaps/src/msg.rs [hunk 2]
```diff
@@ -26,7 +27,35 @@ pub enum FailedDeliveryAction {
...
+    fn json_schema(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        schemars::schema::Schema::from(true)
+    }
+}
+
+impl SerializableJson {
+    pub fn as_value(&self) -> &serde_cw_value::Value {
+        &self.0
+    }
+}
```
Comment (ValarDragon) [inline cosmwasm/contracts/crosschain-swaps/src/msg.rs:56]: Very cool that serde lets you do this

File: cosmwasm/contracts/outpost/src/msg.rs [hunk 3]
```diff
@@ -27,27 +26,25 @@ pub struct Wasm {
...
     pub contract: Addr,
-    pub msg: String,
-}
-
-#[derive(Serialize)]
-struct PrivateCallback {
-    pub contract: String,
-    pub msg: serde_cw_value::Value,
+    pub msg: crosschain_swaps::msg::SerializableJson,
 }
```
Comment (ValarDragon) [inline cosmwasm/contracts/outpost/src/msg.rs:30]: Private Callback was just unused right?

File: tests/ibc-hooks/ibc_middleware_test.go [hunk 4]
```diff
@@ -808,7 +808,7 @@ func (suite *HooksTestSuite) TestBadCrosschainSwapsNextMemoMessages() {
 	recoverAddr := suite.chainA.SenderAccounts[8].SenderAccount.GetAddress()
 	receiver := initializer
 
-	innerMsg := fmt.Sprintf(`{"osmosis_swap":{"swap_amount":"1000","output_denom":"token1","slippage":{"twap": {"window_seconds": 1, "slippage_percentage":"20"}},"receiver":"%s","on_failed_delivery": {"local_recovery_addr": "%s"},"next_memo":%%s}}`,
+	innerMsg := fmt.Sprintf(`{"osmosis_swap":{"swap_amount":"10","output_denom":"token1","slippage":{"twap": {"window_seconds": 1, "slippage_percentage":"20"}},"receiver":"%s","on_failed_delivery": {"local_recovery_addr": "%s"},"next_memo":%%s}}`,
```
Comment (ValarDragon) [inline tests/ibc-hooks/ibc_middleware_test.go:811]:
```suggestion
	// next_memo is set to `%s` after the SprintF. It is then format replaced in each test case.
	innerMsg := fmt.Sprintf(`{"osmosis_swap":{"swap_amount":"10","output_denom":"token1","slippage":{"twap": {"window_seconds": 1, "slippage_percentage":"20"}},"receiver":"%s","on_failed_delivery": {"local_recovery_addr": "%s"},"next_memo":%%s}}`,
```

File: tests/ibc-hooks/ibc_middleware_test.go [hunk 5]
```diff
@@ -820,16 +820,19 @@ func (suite *HooksTestSuite) TestBadCrosschainSwapsNextMemoMessages() {
 		{fmt.Sprintf(innerMsg, `""`), false},
 		{fmt.Sprintf(innerMsg, `null`), true},
 		{fmt.Sprintf(innerMsg, `"{\"ibc_callback\": \"something\"}"`), false},
-		{fmt.Sprintf(innerMsg, `"{\"myKey\": \"myValue\"}"`), true},
-		{fmt.Sprintf(innerMsg, `"{}""`), true},
+		{fmt.Sprintf(innerMsg, `"{\"myKey\": \"myValue\"}"`), false},
```
Comment (ValarDragon) [inline tests/ibc-hooks/ibc_middleware_test.go:823]:
```suggestion
		{fmt.Sprintf(innerMsg, `"{\"myKey\": \"myValue\"}"`), false}, // JSON memo should not be escaped
```

**Review by ValarDragon — APPROVED** [review]

File: cosmwasm/contracts/outpost/src/msg.rs (hunk 3 above)
Comment (nicolaslara) [reply cosmwasm/contracts/outpost/src/msg.rs:30]: It was being used for serializing. It was basically the same but it derived Serialize, which meant we could convert it to json. Now SerializableJson takes care of that

**Pull request merged**